Back to results button in item view

mspalti opened the pull request:
## Description
Adds a "back to results" button to the Item view when the previous url in the router history is a search, browse or recent submissions list.

## Instructions for Reviewers
A button appears in the item view when you navigate from any of the lists mentioned above. Use it to navigate back to the previous page in the pagination list.

The back button is added when the previous url in history begins with `/search`, `/browse`, or `/collections` (see `item.component.ts`). 

Currently, the only way to navigate back to a `page` in the pagination list is via the browser back button. That's a usability problem. The solution here is fairly simple. I spent time looking for an edge case where the approach doesn't work, but I couldn't find one. Logically that makes sense since the back option uses the Router history, but let me know if an edge case exists. 

List of changes in this PR:
* Button added to all item html templates
* Button is shown when `routeService.getPreviousUrl()` returns a matching route
* Clicking on the button uses `Router.navigateByUrl()` to navigate back to the previous pagination list.


## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [ESLint](https://eslint.org/) validation using `yarn lint`
- [x] My PR doesn't introduce circular dependencies (verified via `yarn check-circ-deps`)
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [ ] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [ ] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

mspalti commented:
I refactored the back button to be a shared component and made it theme-able while I was at it.

tdonohue commented on src/app/shared/shared.module.ts:
```diff
@@ -395,6 +398,8 @@ const COMPONENTS = [
   ItemDetailPreviewComponent,
   ItemDetailPreviewFieldComponent,
+  ItemBackButtonComponent,
+  ThemedItemBackButtonComponent,
```
Rather than adding these to the "shared" module, could we add them to the `item-page.module.ts` instead?  It appears they are specific to the ItemPageModule (which is also used by all Entity pages as well), so they don't need to be in the "SharedModule"

mspalti replied:
Sure, I can look into this more. 

We may be able to use the `item-shared-module.ts`. That module is imported by the `entity-group` modules. Also, I just added another wrinkle here. I decided that because the button is now a themed component it makes sense use it for the `browse-by` component back buttons.  If @artlowel agrees with this approach then we would need to add the `ItemSharedModule` to `browse-by.module.ts` as well.

mspalti replied:
The solution I came up with was more complex than I guessed. Because the button is also used for `browse-by` components I decided to make separate modules for both the results back button and the shared browse-by component. That removes both from the "shared" module. 

I know that the Atmire team is working on shrinking the size of `shared.module` and the approach taken here might not align well with what they are thinking. If so, that's fine. I can fix things here if needed.

The back buttons are working now for Admin Search and MyDSpace.

tdonohue reviewed (changes requested):
@mspalti : Great work!  I didn't realize this would be so easy, and I agree that it helps with usability.  Overall, I'm basically a +1 vote, but I have a few minor requests.

1. We need to avoid adding code to the `SharedModule` unless absolutely necessary, as we've found the `SharedModule` has become bloated & we need to shrink it (see #1975 and #1921).  I think this new code can easily move to `ItemPageModule` though (haven't tried it, but it looks like it should cleanly move over)
2. I noticed that this back button does NOT yet appear in the following places.  If possible, it might be nice to enable it in these two situations as well:
    * If you come from the Admin Search results (Login as Admin -> "Admin Search" -> click on a result). 
    * If you come from the MyDSpace list of Submissions or Workflow Tasks (Login -> MyDSpace -> Show="Your Submissions" OR "Workflow Tasks" -> Click "View" button)

Overall, I like this PR & I don't see any immediate issues myself.  I would appreciate it though if @artlowel took a quick look just to make sure he's not seeing any issues with this implementation of a back button.

mspalti commented:
@artlowel, @tdonohue, I decided to use the themed button component with `shared/browse-by/browse-by.component` so there's consistent theming across contexts. Basically, if pagination `Input()`'s are provided  in the back button component the `PaginationService` is used for back navigation. Otherwise we just use the router. 

If this turns out to be a bad idea I can revert back to a separate button for browse-by.

artlowel commented on src/app/entity-groups/journal-entities/item-pages/journal-issue/journal-issue.component.html:
```diff
@@ -1,3 +1,9 @@
+<div class="row" *ngIf="showBackButton | async">
+  <div class="col-12">
+    <ds-themed-results-back-button></ds-themed-results-back-button>
+  </div>
+</div>
+
```
Perhaps you can move this `showBackButton`  check to `ds-results-back-button` itself? that way there'd be less code duplication. Since you're only using a `col-12` (i.e. a full width `col`) you should be able to just leave out the `row` and `col` divs without seeing a difference

artlowel commented on src/app/entity-groups/journal-entities/item-pages/journal-issue/journal-issue.component.scss:
```diff
@@ -1 +1,5 @@
 @import '../../../../../styles/variables.scss';
+
+button.back-btn {
+  margin-bottom: 10px;
```
Perhaps use [a bootstrap `m-` class](https://getbootstrap.com/docs/4.1/utilities/spacing/) rather than setting the margin manually in pixels, that would scale with the theme.
 
And again, if you do that in `ds-results-back-button` you don't have to repeat it for every entity type

artlowel reviewed (changes requested):
Thanks @mspalti!

It works, and I like the look and the positioning. I just have a few suggestions to reduce code duplication.

I appreciate that you've made it themeable out of the box!

artlowel commented on src/app/shared/browse-by/browse-by.component.html:
```diff
@@ -3,7 +3,7 @@ <h3 [ngClass]="{'sr-only': parentname }">{{title | translate}}</h3>
   <div *ngIf="objects?.hasSucceeded && !objects?.isLoading && objects?.payload?.page.length > 0" @fadeIn>
       <div *ngIf="shouldDisplayResetButton$ |async" class="mb-2 reset">
-          <button class="btn btn-secondary" (click)="back()"><i class="fas fa-arrow-left"></i> {{'browse.back.all-results' | translate}}</button>
+        <ds-results-back-button [paginationConfig]="paginationConfig" [previousPage$]="previousPage$"></ds-results-back-button>
```
Perhaps use the themed version here as well?

artlowel commented on src/app/shared/browse-by/browse-by.component.html:
```diff
@@ -18,7 +18,7 @@ <h3 [ngClass]="{'sr-only': parentname }">{{title | translate}}</h3>
   <div *ngIf="!objects?.isLoading && objects?.payload?.page.length === 0">
     <div *ngIf="shouldDisplayResetButton$ |async" class="d-inline-block mb-4 reset">
-      <button class="btn btn-secondary" (click)="back()"><i class="fas fa-arrow-left"></i> {{'browse.back.all-results' | translate}}</button>
+      <ds-results-back-button [paginationConfig]="paginationConfig" [previousPage$]="previousPage$" ></ds-results-back-button>
```
… and here

mspalti commented:
Thanks for the great suggestions, @artlowel. I think all have been addressed.

artlowel reviewed (approved):
Thanks for taking them!

One thing that could still be improved is when you start from a metadata browse page, e.g. browse by author:
- click on an author name 
- then click on an item 
- then on the new back button

Now you're back on the browse page for that particular author, but there's no back button to take you to the initial browse page.

However that doesn't necessarily need to be addressed in this PR, or perhaps even at all, as that will likely make the logic a lot more complicated, introduce more edge cases etc.

So this PR can me merged as-is for me

tdonohue reviewed (commented):
@mspalti : I tested this, and it works well overall.  But, I think what @artlowel points out is a **small bug**.  There's now an (accidental) change in behavior in the "Browse by Author" & "Browse by Subject" pages.

On `main`, if you click "Browse by Author" and then select an Author's name: you always see a button to return to "<- All browse results".  Here's an example from the demo site:
![main-behavior](https://user-images.githubusercontent.com/483997/213511228-3e307f2a-c476-4c36-8b64-d8d47be014fb.png)

But, now that "All browse results" button disappears in the scenario that @artlowel points out.  This also occurs in "Browse by Subject".

1. Click on Browse by Author/Subject
2. Click an Author name or Subject term
3. You will see the "Back to Results" button at this point on the "Browsing by..." page
4. Now, click an Item in the list.  Now the "Back to Results" button moves to the item page.  Click it.
5. You go back to the "Browsing by..." page (from step 3), but the "Back to Results" button is gone now.  So, there is no way to get back to the overall "Browse by Author" or "Browse by Subject" page (from step 1) unless you click the option in the header again.

I realize this might not be easy to resolve in this PR without more work.  But, I wonder if maybe there'd be a way to change the breadcrumbs to say something like:
`Home > Browse by Author > Browsing by Author "[name]"`

That way, there's some way to get back to the overall "Browse by Author" page.  Otherwise, I worry that users will see this new behavior as a bug.... there is no way to get back to the "Browse by" option where they started.

I won't officially block this PR, but I think it's worth looking at whether there's a way to avoid this new bug.

mspalti commented:
@tdonohue, @artlowel, sorry about the latest bug. I didn't notice it after the last code change. 

When I moved the conditional logic from the item page components into the button component that caused the scenario you noticed. The browse component has it's own show/hide logic that's inconsistent with the item logic we moved into the button. I decided the best way to handle this is to add a new `alwaysShowButton` input.  This fixed the problem with the browse button and might be useful in other future situations.

I also noticed a silly problem in the theme component that's now fixed. _I think_ everything is working now. Check me on that. :-)

mspalti commented:
I made one last change/improvement. The back button component shouldn't worry about setting up a `back()` function  that works for different parent components. That's a little confusing and error-prone.  It's bothered me for a while but I haven't had time to consider the options.

Here's the change: functionally everything's the same, but the `back()`  function is now defined in the parents (`browse-by.component` and `item.component` respectively) and passed as an input to `results-back-button.component`. That simplifies the button component and helps avoid the kind of problem we ran into above, where the button was being accidentally hidden.

mspalti commented:
I've seen a somewhat intermittent test error in my local environment that never seemed to manifest in PR tests. That changed today and I was able to track it down. It was caused by an unnecessary module import in one of the tests. The problem didn't affect builds and running the code.

mspalti commented:
Just to wrap up. (I'm optimistically assuming this work is complete.) 

The back button feature in this PR works right now with items and metadata browse lists. Although given the way it's implemented it could be used in other situations as well, perhaps navigating between entities types (if that's possible). 

To use the themed button component, a parent component just needs to decide when to show the button, provide a label translation (or accept the default "back to results" label), and create a `back()` method that the button will use for navigation. The label and `back()` method are inputs to the button component.

tdonohue commented on src/app/item-page/simple/item-types/shared/item.component.ts:
```diff
@@ -38,12 +45,33 @@ export class ItemComponent implements OnInit {
     this.itemPageRoute = getItemPageRoute(this.object);
+    // hide/show the back button
+    this.showBackButton = this.routeService.getPreviousUrl().pipe(
+      filter(url => /^(\/search|\/browse|\/collections|\/admin\/search|\/mydspace)/.test(url)),
```
We might think about turning this RegEx into a constant at the top of this component (to make it easier to find).  I suspect it will need to change for some (highly custom) DSpace sites.. or even in the future for DSpace itself.  While I don't think it needs to be made configurable yet, it could be nice to simply make it easier to find/update at the component level.

mspalti replied:
That's a good suggestion, @tdonohue . It's done.

tdonohue reviewed (approved):
👍 Thanks @mspalti !  I retested this today, and verified that the small bug that I saw (and @artlowel originally reported) is now fixed!  I'm not finding any other immediately obvious issues, so I think this looks good (and if we find later issues we can fix them later).

I have one very minor comment inline.  Nothing required, but a minor code cleanup that could make the path Regex you are using easier to update/find.

artlowel reviewed (approved):
Thanks @mspalti!

I've tested it again, and indeed the issue is gone. I also agree with the reasoning to move the showBackButton code back

tdonohue commented:
Merging as this is now at +2.  Thanks again @mspalti !
